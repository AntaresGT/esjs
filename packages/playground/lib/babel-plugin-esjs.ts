import type { ParserOptions } from '@babel/parser'
import { parse } from '@babel/parser'
import esjs from '@esvue/esjs/dist/index.js'

export default function EsJs() {
  return {
    name: 'esjs',
    parserOverride(code: string, options?: ParserOptions) {
      const output = esjs.transpile(code)
      return parse(output, options)
    },
  }
}
