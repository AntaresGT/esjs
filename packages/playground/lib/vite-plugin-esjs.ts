import * as babel from '@babel/core'
import type { Plugin } from 'vite'
import EsJs from './babel-plugin-esjs'

export default (): Plugin => ({
  name: 'vite-plugin-esjs',
  enforce: 'pre',
  transform(raw, id) {
    if (!/\.esjs$/.test(id))
      return

    const result = babel.transformSync(raw, {
      babelrc: false,
      ast: true,
      plugins: [
        EsJs(),
      ],
      sourceFileName: id,
      configFile: false,
    })

    return `
<script setup lang="ts">
import { Terminal } from "esvue";
import { usarConsola } from "esvue";

const consola = usarConsola();

(async () => {
  ${result.code}
})();
</script>

<template>
<Terminal />
</template>`
  },
})
