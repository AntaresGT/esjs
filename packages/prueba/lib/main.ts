import logSymbols from 'log-symbols'
import { isEqual } from './isEqual'

interface Resultado {
  numeroPruebas: number
  exitosas: number
  fallidas: number
}

class PruebaError extends Error {
  constructor(public pruebaNombre: string, public error: Error, public dontWarn = true) {
    super(error.message)
  }
}

class PruebasError extends Error {
  constructor(public dontWarn = true) {
    super('Se encontraron errores')
  }
}

export function pruebas(pruebas: any) {
  let failures = 0

  for (const pruebaNombre in pruebas) {
    const pruebaFuncion = pruebas[pruebaNombre]

    try {
      prueba(pruebaNombre, pruebaFuncion)
    }
    catch (e) {
      failures++
    }
  }

  const result = obtenerResultado(pruebas, failures)

  // eslint-disable-next-line no-console
  console.log(obtenerResumen(result))

  if (failures > 0)
    throw new PruebasError()

  return result
}

export function prueba(pruebaNombre: string, pruebaFuncion: () => void) {
  try {
    pruebaFuncion()

    // eslint-disable-next-line no-console
    console.log(`%c${logSymbols.success} ${pruebaNombre}`, 'color: #14b8a6; font-size: 14px; padding: 2px 4px;')
  }
  catch (error: any) {
    // eslint-disable-next-line no-console
    console.groupCollapsed(`%c${logSymbols.error} ${pruebaNombre}`, 'color: #f43f5e; font-size: 14px; padding: 2px 4px;')
    console.error(error?.stack)
    // eslint-disable-next-line no-console
    console.groupEnd()

    throw new PruebaError(pruebaNombre, error)
  }
}

export function afirmar(valor: boolean, mensaje?: string) {
  if (!valor) {
    mensaje = mensaje ?? 'afirmar(): Se esperaba "verdadero", pero se recibi贸 "falso"'
    throw new Error(mensaje)
  }
}

export const assert = afirmar

export function afirmarIguales(esperado: any, actual: any, mensaje?: string) {
  return afirmar(isEqual(esperado, actual, true), mensaje ?? `afirmarIguales(): "${esperado}" !== "${actual}"`)
}

export function afirmarSimilares(esperado: any, actual: any, mensaje?: string) {
  // eslint-disable-next-line eqeqeq
  if (esperado != actual) {
    mensaje = mensaje ?? `afirmarSimilares(): "${esperado}" != "${actual}"`
    throw new Error(mensaje)
  }
}

export function afirmarMatricesIguales(matrizEsperada: Array<any>, matrizActual: Array<any>) {
  return afirmarIguales(matrizEsperada, matrizActual, `afirmarMatricesIguales(): "${matrizEsperada}" != "${matrizActual}"`)
}

export function afirmarObjetosIguales(objetoEsperado: any, objetoActual: any) {
  return afirmarIguales(objetoEsperado, objetoActual, `afirmarObjetosIguales(): "${objetoEsperado}" != "${objetoActual}"`)
}

export function afirmarMatricesSimilares(matrizEsperada: Array<any>, matrizActual: Array<any>, mensaje?: string) {
  if (!isEqual(matrizEsperada, matrizActual, false)) {
    mensaje = mensaje ?? `afirmarMatricesSimilares(): "${matrizEsperada}" != "${matrizActual}"`
    throw new Error(mensaje)
  }
}

export function afirmarObjetosSimilares(objetoEsperado: any, objetoActual: any, mensaje?: string) {
  if (!isEqual(objetoEsperado, objetoActual, false)) {
    mensaje = mensaje ?? `afirmarObjetosSimilares(): "${objetoEsperado}" != "${objetoActual}"`
    throw new Error(mensaje)
  }
}

export function obtenerResumen(result: Resultado) {
  const executedTestsString = result.numeroPruebas === 1 ? 'Se ejecut贸 1 prueba' : `Se ejecutaron ${result.numeroPruebas} pruebas`
  const numberOfSuccessesString = result.exitosas === 1 ? '1 exitosa' : `${result.exitosas} exitosas`
  const numberOfFailuresString = result.fallidas === 1 ? '1 fallida' : `${result.fallidas} fallidas`

  return `${executedTestsString}: \n ${numberOfSuccessesString} \n ${numberOfFailuresString}`
}

export function obtenerResultado(pruebas: any, fallidas: number): Resultado {
  const numeroPruebas = Object.keys(pruebas).length
  const exitosas = numeroPruebas - fallidas

  return {
    numeroPruebas,
    exitosas,
    fallidas,
  } as Resultado
}

export function afirmarVerdadero(valor: boolean, mensaje?: string) {
  return afirmarIguales(true, valor, mensaje ?? 'afirmarVerdadero(): Se esperaba "verdadero", pero se recibi贸 "falso"')
}

export function afirmarFalso(valor: boolean, mensaje?: string) {
  return afirmarIguales(false, valor, mensaje ?? 'afirmarFalso(): Se esperaba "falso", pero se recibi贸 "verdadero"')
}

export function afirmarDistinto(esperado: any, actual: any, mensaje?: string) {
  return afirmar(!isEqual(esperado, actual, true), mensaje ?? `afirmarDistinto(): "${esperado}" === "${actual}"`)
}
