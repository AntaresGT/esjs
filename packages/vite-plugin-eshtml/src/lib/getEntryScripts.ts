/**
 * Get entry scripts from HTML.
 */
export function getEntryScripts(html: string): Map<string, string> {
  const matches = html.match(/<script.*?src="(.+?\.esjs)".*?><\/script>/g)

  if (!matches) {
    return new Map()
  }

  const entryScripts = new Map<string, string>()

  for (const match of matches) {
    const src = match.match(/<script.*?src="(.+?\.esjs)".*?><\/script>/)?.[1]

    if (src) {
      entryScripts.set(src.replace('.esjs', '.js'), src)
    }
  }

  return entryScripts
}
