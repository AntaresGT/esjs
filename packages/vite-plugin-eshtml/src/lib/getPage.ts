import { Opciones, Pagina } from '../index'

export function getPage(
  options: Opciones,
  pageName: string,
): Pagina | undefined {
  if (pageName === 'index') {
    return options?.paginas?.indice ?? options?.paginas?.index
  }

  return options?.paginas?.[pageName]
}
