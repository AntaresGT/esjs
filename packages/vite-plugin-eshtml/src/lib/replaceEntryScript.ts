/**
 * Replace EsJS entry script with a JS script that will load the EsJS script.
 */
export function replaceEntryScript(
  html: string,
  scriptType: string = 'module',
): string {
  return html.replace(/<script.*?src="(.+?\.esjs)".*?><\/script>/g, (m, p1) => {
    return `<script src="${p1.replace(
      '.esjs',
      '.js',
    )}" type="${scriptType}"></script>`
  })
}
