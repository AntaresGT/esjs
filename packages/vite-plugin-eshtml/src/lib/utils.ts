import { compile as EsHTMLCompile } from '@es-js/eshtml'

export function compile(eshtml: string): string {
	return EsHTMLCompile(eshtml, { from: 'eshtml', to: 'html' })
}

export function dfs(keys: string[], value: any, res: Record<string, any>) {
	if (keys.length) {
		const strItem = keys.shift()
		if (!keys.length) {
			res[strItem!] = value
		} else {
			const tmp = res[strItem!] ? res[strItem!] : (res[strItem!] = {})
			dfs(keys, value, tmp)
		}
	}
	return res
}

export function dfs2(
	rebuildData: Record<string, any>,
	key: string,
	value: Record<string, any>,
) {
	const tmp = rebuildData[key] ? rebuildData[key] : (rebuildData[key] = {})
	if (Object.prototype.toString.call(value).slice(8, -1) === 'Object') {
		const nextKey = Object.keys(value)[0]
		dfs2(tmp, nextKey, value[nextKey])
	} else {
		rebuildData[key] = value
	}
}
