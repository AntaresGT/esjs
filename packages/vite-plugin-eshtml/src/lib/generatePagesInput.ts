import { Opciones } from '../index'
import { getPage } from './getPage'

export function generatePagesInput(
	input: any,
	options: Opciones,
): Record<string, string> {
	const newInput: Record<string, string> = { ...input }

	for (const pageName of Object.keys(options?.paginas || {})) {
		const page = getPage(options, pageName)
		if (page && !newInput[pageName]) {
			const pagePath = `${pageName}.html`
			newInput[pageName] = pagePath
		}
	}

	return newInput
}
