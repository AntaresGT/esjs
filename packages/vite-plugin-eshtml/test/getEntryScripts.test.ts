// import vite from 'vite'
import { expect, it } from 'vitest'
import { getEntryScripts } from '../src/lib/getEntryScripts'

it('getEntryScripts', async () => {
	const html = `<!DOCTYPE html>
<html lang="es">

  <head>
    <meta charset="UTF-8">
    <link rel="icon" href="/favicon.ico">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>EsHTML</title>
  </head>

  <body>
    <script src="/fuente/indice.esjs" type="module"></script>
    <script src="/fuente/utilidades.esjs" type="module"></script>
  </body>

</html>`

	const output = getEntryScripts(html)

	expect(output).toMatchSnapshot()
})
