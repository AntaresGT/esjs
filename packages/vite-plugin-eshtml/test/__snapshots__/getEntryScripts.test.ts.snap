// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getEntryScripts 1`] = `
Map {
  "/fuente/indice.js" => "/fuente/indice.esjs",
  "/fuente/utilidades.js" => "/fuente/utilidades.esjs",
}
`;
